File: 移除链表元素.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */


struct ListNode* removeElements(struct ListNode* head, int val){
    struct ListNode *p,*q,*ppre;
    ppre=p=head;
    if(p==NULL)
        return 0;
    else{
        while(p->next!=NULL){
            if(p->val==val&&p==head){
                q=p;
                p=p->next;
                ppre=head=p;
                free(q);
            }  
            else
                if(p->val==val&&p!=head){
                    q=p;
                    ppre->next=q->next;
                    p=q->next;
                    free(q);
                }
                else{
                    ppre=p;
                    p=p->next;
                }
        }
        if(p->val==val&&p==head)
            return 0;
        else
            if(p->val==val){
                q=p;
                ppre->next=q->next;
                free(q);
                return head;
            }
            else
                return head;
    }
}

File: 2020年11月/移除链表元素.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def removeElements(self, head, val):
        """
        :type head: ListNode
        :type val: int
        :rtype: ListNode
        """
        if head==None: #判断头是否为空，头为空则返回空
            return None
        node=head
        while node.next!=None: #去除中间与val相等的元素
            if node.next.val==val:
                node.next=node.next.next
            else:
                node=node.next
        if head.val==val: #判断头是否与val相同
            head=head.next
        return head